### ISSUE TITLE
Deprecate `npm search` in npm 4.0

### ISSUE THREAD

**Issue by smikes** [npm/npm#11035]
TL;DR:

Next version of npm@3, start warning that `npm search` will be deprecated as of 4.0

Remove the command in npm@4

Consider re-introducing `npm search` in >=5 as a command that queries the `npmjs.org` search API endpoint.

Motivations:

`npm search` as an offline command has been broken for >2 years.  People still try to use it though.
Offline-supporting alternatives (npm-kludge-search; there may be others) are still slow and awkward, unlikely to be integrated into npm mainline anytime soon.
Eventual goal is to have `npm search` provide the same results as the npmjs.org search endpoint, but implementing this feature a) requires www/registry coordination and b) is not even visible from the current  CLI roadmap

Therefore:

Put CLI `npm search` on the deprecation track now and revive it when the www/registry coordination can be done.

Relevant search issues: #11010 #10615 #9251 #4360 #8700 #6016 #9259 #5144 , many others

**Comment by zkat** [comment]
This item is [already in the Roadmap](https://github.com/npm/npm/wiki/Roadmap-area-of-focus%3A-search), but I'm leaving this particular issue open for now as a concrete proposal of how to handle it.

**Comment by othiym23** [comment]
As a team, we agree that the current situation is unacceptable. That said, deprecating search is an even bigger hammer than the status quo. We'd like to at least make search work reliably within the CLI, even if it's slow. That is to say, the scope of work we're interested in here is making it not crash because the process is out of memory. @aredridel has put some work into making it stream the JSON when searching, which should address search crashing, and I think that's probably our bridge strategy until as a company we have the resources to tackle making search nicer.

As always, thank for the detailed and thoughtful writeup, Sam! It's much appreciated!

**Comment by iamjochem** [comment]
could SQLite be a way out? I hacked together a proof-of-concept which someone might want to take a look at:

>  https://github.com/iamjochem/npm-search-sqlite

it's bare bones but maybe it's something that could be used as a starting point?
